Source/DafnyLanguageServer/Handlers/DafnyWorkspaceSymbolHandler.cs

```diff
@@ -0,0 +1,104 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.Workspace;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using OmniSharp.Extensions.LanguageServer.Protocol.Workspace;
+
+namespace Microsoft.Dafny.LanguageServer.Handlers {
+  public class DafnyWorkspaceSymbolHandler : WorkspaceSymbolsHandlerBase {
+    private readonly ILogger logger;
+    private readonly IProjectDatabase projects;
+    private readonly DafnyOptions dafnyOptions;
+
+    public DafnyWorkspaceSymbolHandler(ILogger<DafnyWorkspaceSymbolHandler> logger, IProjectDatabase projects, DafnyOptions dafnyOptions) {
+      this.logger = logger;
+      this.projects = projects;
+      this.dafnyOptions = dafnyOptions;
+    }
+
+    protected override WorkspaceSymbolRegistrationOptions CreateRegistrationOptions(WorkspaceSymbolCapability capability,
+      ClientCapabilities clientCapabilities) {
+      return new WorkspaceSymbolRegistrationOptions {
+        WorkDoneProgress = false
+      };
+    }
+
+    public override async Task<Container<SymbolInformation>?> Handle(WorkspaceSymbolParams request, CancellationToken cancellationToken) {
+      var queryText = request.Query.ToLower();
+      Dictionary<SymbolInformation, int> result = new Dictionary<SymbolInformation, int>();
+      // Using a LINQ traversal here would be more readable, but to be as responsive
+      // to cancellations as possible, we instead use a loop:
+      foreach (var projManager in projects.Managers) {
+        cancellationToken.ThrowIfCancellationRequested();
+        var state = await projManager.GetStateAfterParsingAsync();
+        foreach (var def in state.SymbolTable.Definitions) {
+          cancellationToken.ThrowIfCancellationRequested();
+          // CreateLocation called below uses the .Uri field of Tokens which
+          // seems to occasionally be null while testing this from VSCode
+          if (def.NameToken == null || def.NameToken.Uri == null) {
+            logger.LogWarning($"Definition without name token in symbol table: {def}");
+            continue;
+          }
+          // GetDescription is not quite the right thing to use here,
+          // as it includes things like keywords and the signature of the
+          // the definitions. However, extracting just the name from
+          // an ISymbol requires explicit case splitting for each definition
+          var description = def.GetDescription(dafnyOptions);
```

Comment:
Wouldn't `def.NameToken.val` give you the name?

Also, this doesn't match on qualified names, does it? (could be a future feature)

Reply:
Oh, I must have missed that. I'll change it. Good callout about the qualified names. I'll leave a note here about it.